src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -272,64 +281,92 @@ public override async Task<VSCodeAction> ResolveCodeActionsAsync(VSCodeAction co
 
         [Obsolete]
 #pragma warning disable CS0809 // Obsolete member overrides non-obsolete member
-        public override async Task<ProvideSemanticTokensResponse> ProvideSemanticTokensAsync(SemanticTokensParams semanticTokensParams, CancellationToken cancellationToken)
+        public override async Task<ProvideSemanticTokensResponse> ProvideSemanticTokensAsync(
+            ProvideSemanticTokensParams semanticTokensParams,
+            CancellationToken cancellationToken)
 #pragma warning restore CS0809 // Obsolete member overrides non-obsolete member
         {
             if (semanticTokensParams is null)
             {
                 throw new ArgumentNullException(nameof(semanticTokensParams));
             }
 
-            var csharpDoc = GetCSharpDocumentSnapshsot(semanticTokensParams.TextDocument.Uri);
+            var csharpDoc = GetCSharpDocumentSnapshsot(semanticTokensParams.TextDocument.Uri.ToUri());
             if (csharpDoc is null)
             {
                 return null;
             }
 
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync((int)semanticTokensParams.RequiredHostDocumentVersion, csharpDoc, cancellationToken);
```

Comment:
What does this line/the document synchronizer do?

Reply:
I'm a bit fuzzy on the specifics of the `TrySynchronizeVirtualDocumentAsync` internals myself since it eventually starts doing callbacks, but here's my general understanding:

On the VS client we call `TrySynchronizeVirutalDocumentAsync` with a `requiredHostDocumentVersion` given to us by the Razor LSP server, reflecting the version of the document that we're trying to get any dependent servers to agree on. If they already match great it basically no-ops, otherwise we wait on the `[GetSynchronizationTaskAsync](https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultLSPDocumentSynchronizer.cs#L68)` which resolves once the DocumentContext [has seen the appropriate document version](https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultLSPDocumentSynchronizer.cs#L191) come across the `documentManager.Changed` Event.

Thankfully we already use it [in a couple places](https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentPullDiagnosticsHandler.cs#L96-L99) for a similar reason so I'm fairly confident on using it this way.

Reply:
Makes sense, appreciate the explanation!